File: 07-Encordados-JPA-Hibernate-Java/src/main/java/repositories/raquetas/RaquetasRepositoryImpl.java
package repositories.raquetas;

import db.HibernateManager;
import exceptions.RaquetaException;
import exceptions.TenistaException;
import jakarta.persistence.TypedQuery;
import models.Raqueta;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.logging.Logger;

public class RaquetasRepositoryImpl implements RaquetasRepository {
    private final Logger logger = Logger.getLogger(RaquetasRepositoryImpl.class.getName());

    @Override
    public List<Raqueta> findAll() {
        logger.info("findAll()");
        HibernateManager hb = HibernateManager.getInstance();
        hb.open();
        TypedQuery<Raqueta> query = hb.getManager().createNamedQuery("Raqueta.findAll", Raqueta.class);
        List<Raqueta> list = query.getResultList();
        hb.close();
        return list;
    }

    @Override
    public Optional<Raqueta> findById(UUID uuid) {
        logger.info("findById()");
        HibernateManager hb = HibernateManager.getInstance();
        hb.open();
        Optional<Raqueta> raqueta = Optional.ofNullable(hb.getManager().find(Raqueta.class, uuid));
        hb.close();
        return raqueta;
    }

    @Override
    public Raqueta save(Raqueta entity) {
        logger.info("save()");
        HibernateManager hb = HibernateManager.getInstance();
        hb.open();
        hb.getTransaction().begin();

        try {
            hb.getManager().merge(entity);
            hb.getTransaction().commit();
            hb.close();
            return entity;

        } catch (Exception e) {
            throw new RaquetaException("Error al salvar raqueta con uuid: " + entity.getUuid() + "\n" + e.getMessage());
        } finally {
            if (hb.getTransaction().isActive()) {
                hb.getTransaction().rollback();
            }
        }
    }

    @Override
    public Boolean delete(Raqueta entity) {
        logger.info("delete()");
        HibernateManager hb = HibernateManager.getInstance();
        hb.open();
        try {
            hb.getTransaction().begin();
            // Ojo que borrar implica que estemos en la misma sesión y nos puede dar problemas, por eso lo recuperamos otra vez
            entity = hb.getManager().find(Raqueta.class, entity.getUuid());
            hb.getManager().remove(entity);
            hb.getTransaction().commit();
            hb.close();
            return true;
        } catch (Exception e) {
            throw new TenistaException("Error al eliminar tenista con uuid: " + entity.getUuid() + " - " + e.getMessage());
        } finally {
            if (hb.getTransaction().isActive()) {
                hb.getTransaction().rollback();
            }
        }
    }
}

File: 08-Encordados-JPA-Hibernate-Kotlin/src/main/kotlin/repositories/raqueta/RaquetasRepositoryImpl.kt
package repositories.raqueta

import db.HibernateManager
import db.HibernateManager.manager
import models.Raqueta
import mu.KotlinLogging
import java.util.*
import javax.persistence.TypedQuery


private val logger = KotlinLogging.logger {}

class RaquetasRepositoryImpl : RaquetasRepository {

    override fun findAll(): List<Raqueta> {
        logger.debug { "findAll()" }
        var raquetas = mutableListOf<Raqueta>()
        HibernateManager.query {
            val query: TypedQuery<Raqueta> = manager.createNamedQuery("Raqueta.findAll", Raqueta::class.java)
            raquetas = query.resultList
        }
        return raquetas
    }

    override fun findById(id: UUID): Raqueta? {
        logger.debug { "findById($id)" }
        var raqueta: Raqueta? = null
        HibernateManager.query {
            raqueta = manager.find(Raqueta::class.java, id)
        }
        return raqueta
    }

    fun findByMarca(marca: String): List<Raqueta> {
        logger.debug { "findByMarca($marca)" }
        var raquetas = mutableListOf<Raqueta>()
        HibernateManager.query {
            // Podemos definir la consulta sobre la marcha
            val query: TypedQuery<Raqueta> =
                manager.createQuery(
                    "SELECT r FROM Raqueta r WHERE r.marca = :marca",
                    Raqueta::class.java
                )
            query.setParameter("marca", marca)
            raquetas = query.resultList
        }
        return raquetas
    }

    override fun save(entity: Raqueta): Raqueta {
        logger.debug { "save($entity)" }
        HibernateManager.transaction {
            manager.merge(entity)
        }
        return entity
    }

    override fun delete(entity: Raqueta): Boolean {
        var result = false
        logger.debug { "delete($entity)" }
        HibernateManager.transaction {
            // Ojo que borrar implica que estemos en la misma sesión y nos puede dar problemas, por eso lo recuperamos otra vez
            val raqueta = manager.find(Raqueta::class.java, entity.uuid)
            raqueta?.let {
                manager.remove(it)
                result = true
            }
        }
        return result
    }
}

